### PR TITLE
Enums

### PR DESCRIPTION
**Pull request by wajahat414**
All functions were tested continously while developing
linter errors were removed
comments added for testing purposes

### CODE REVIEW

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUIRED** :recycle: :recycle: 

**Good effort and progress so far. Please keep it going.** :ok_hand: :ok_hand: :+1: 

Your enumerable methods are working but we need to cover all the edge cases a typical Ruby enumerable method(just like the ones you have created) would cover.

Firstly,
- [x] Please remove all commented code as well as all other codes not in your enumerable methods.

Below are the remaining cases your enumerable methods are not covering: 

**my_each**

- [x] It should return an Enumerator if no block is given.
- [x] When a block is given, when self is an array, it should return the array itself after calling the given block once for each element in self.
- [x] When a block is given, when self is a range, it should return the range itself after calling the given block once for each element in self.
- [x] When a block is given, when self is a hash, it behaves like #each, calls the given block once for each element in self.

**my_each_with_index**

- [x] It should return an enumerator if no block is given.
- [x] When a block is given, when self is an array, it should return the array itself after calling the given block once for each element in self.
- [x] When a block is given, when self is a range, it should return the range itself after calling the given block once for each element in self.
- [x] When a block is given, when self is a range, it behaves like #each_with_index, calls the given block once for each element in self.
- [x] When a block is given, when self is a hash, it behaves like #each_with_index, calls the given block once for each element in self.

**my_select**

- [x] It should return an enumerator if no block is given.

**my_all?**

- [x] When no block or argument is given, it should return false when one of the collection members are false or nil.
- [ ] When a class is passed as an argument, it should return false if any of the collection is not a member of such class.
- [ ]  When a Regex is passed as an argument, it should return false if any of the collection does not match the Regex.
- [ ] When a pattern other than Regex or a Class is given, it should return false if any of the collection does not match the pattern.

**my_any?**

- [ ] It should return false if the block ever returns a value false or nil.
- [ ] It should return false if the block ever returns a value false or nil::range.
- [ ] When a Regex is passed as an argument, it should return true if any of the collection matches the Regex.
- [ ] When a Regex is passed as an argument, it should return false if none of the collection matches the Regex.
- [ ] When a pattern other than Regex or a Class is given, it should return false if none of the collection matches the pattern.
- [ ] When a pattern other than Regex or a Class is given, it should return true if any of the collection matches the pattern.

**my_none?**

- [ ] When a Regex is passed as an argument, it should return true if none of the collection matches the Regex.
- [ ]  When a Regex is passed as an argument, it should return false if any of the collection matches the Regex.
- [ ]  When a pattern other than Regex or a Class is given, it should return true only if none of the collection matches the pattern.
- [ ] When a pattern other than Regex or a Class is given, it should return false only if one of the collection matches the pattern.

**my_count**

- [ ] It should return the number of items in enum through enumeration::range

**my_map**

- [ ] It should return a new array with the results of running a given block once for every element in enum.
- [ ] It should return an Enumerator if no block is given.
- [ ] It  accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
- [ ] It executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
- [ ] It should not mutate the original array.

**my_inject**

- [ ]  When a block is given without an initial value, it combines all elements of enum by applying a binary operation, specified by a block::range.
- [ ] When a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method.
- [ ] wWen a symbol is specified without an initial value, it should combine each element of the collection by applying the symbol as a named method:range.
- [ ] When a symbol is specified with an initial value, it should combine each element of the collection by applying the symbol as a named method.
- [ ] When a symbol is specified with an initial value, it should combine each element of the collection by applying the symbol as a named method::range.

**Please incorporate these behaviors into your enumerable methods.** :hammer_and_pick: :hammer: 

**Submit another CRR when you're ready.** 

**Reach out to me on slack if require further clarification.**

**Happy coding** :computer: :computer_mouse:

File: enums.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    while index < to_a.length
+      if instance_of?(Hash)
+        yield(keys[index], values[index])
+      else
+        yield(to_a[index])
+      end
+      index += 1
+    end
+    self
+  end
```
Comment (Berabjesus) [inline enums.rb]: - [x]  when a block is given when self is a hash behaves like #each calls the given block once for each element in self. 
```
  enum = { a: 1, b: 2, c: 3, d: 4, e: 5 }
  my_each_output = ''
  block = proc { |num| my_each_output += num.to_s }
  enum.each(&block)
  each_output = my_each_output.dup
  my_each_output = ''
  enum.my_each(&block)
  p my_each_output # "abcde"
  p each_output  # "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
```

File: enums.rb [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    return to_enum unless block_given?
+
+    my_each do |k|
+      if yield(k)
+        arr[i] = k
+        index += 1
+      end
+    end
+    arr
+  end
```
Comment (Berabjesus) [inline enums.rb]: - [x]  returns an array containing all elements of enum for which the given block returns a true value

```
  array = [1,2,3,4,5,6]
  block = proc { |num| num < (0 + 9) / 2 }
  p array.select(&block) # [1, 2, 3]
  p array.my_select(&block) # Error
```

- [x] does not mutate the original array

```
  array.my_select { |num| num + 1 } # Error
```

File: enums.rb [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      return false unless k
+
+      if block_given?
+        return false unless yield(k)
+      elsif (flag = [Integer, Float, String, Numeric].include?(pattern)) || pattern.to_s.include?('?')
+        return false unless check?(k, pattern, flag)
+      end
+    end
+    true
+  end
```
Comment (Berabjesus) [inline enums.rb]: - [x] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

```
  p [3,3,3,3,4].all?(3) # false
  p [3,3,3,3,4].my_all?(3) # true
```

File: enums.rb [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      if block_given?
+        return true if yield(k)
+      elsif (flag = [Integer, Float, String, Numeric].include?(pattern)) || pattern.to_s.include?('?')
+        return true if check?(k, pattern, flag)
+      elsif k
+        return true
+      end
+    end
+    false
+  end
```
Comment (Berabjesus) [inline enums.rb]: - [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

```
  p %w[dog door rod blade].any?('test') # false
  p %w[dog door rod blade].my_any?('test') # true
```

File: enums.rb [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      if block_given?
+        return false if yield(k)
+      elsif (flag = [Integer, Float, String, Numeric].include?(pattern)) || pattern.to_s.include?('?')
+        return false if check?(k, pattern, flag)
+      elsif k
+        return false
+      end
+    end
+    true
+  end
```
Comment (Berabjesus) [inline enums.rb]: - [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

```
  p %w[dog door rod blade].none?(5) # true
  p %w[dog door rod blade].my_none?(5) # false
```

File: enums.rb [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        mem = mem.public_send(vos, arr[i + 1]) if i < arr.to_a.length - 1
+      elsif ops
+        mem = mem.public_send(ops, arr.to_a[i + 1]) if i < arr.to_a.length - 1
+
+      end
+      mem = yield(mem, k) if block_given?
+    end
+    mem = yield(mem, vos) if block_given? && vos
+    mem
+  end
```
Comment (Berabjesus) [inline enums.rb]: - [x] raises a "LocalJumpError" when no block or argument is given
- [x] when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block.
```
  operation =  proc { |sum, n| sum + n } 
  p [1,2,3,4,5,6].inject(&operation) # 21
  p [1,2,3,4,5,6].my_inject(&operation) # 22
```

- [ ]  when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range

```
  p range.my_inject { |prod, n| prod * n } 
# 6336269417023620425752626701263493509245341993533440000000000000

  p range.inject { |prod, n| prod * n } 
# 1267253883404724085150525340252698701849068398706688000000000000
```

- [ ]  when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

```
  p range.my_inject(2, :*) # 1267253883404724085150525340252698701849068398706688000000000000
  p range.inject(2, :*) # 2534507766809448170301050680505397403698136797413376000000000000
```

**Review by Berabjesus — CHANGES_REQUESTED** [review]
## Status: CHANGES REQUIRED :warning: :construction: 
Hi @wajahat414 greater job on working on the previous comments 👏 

#### However, you do have a few changes to carry out before we proceed so please take a look at the things you have to work on and their respective review comments below. 

#### ❗ Please have a look at the inline comments below for each test case.

### my_each
- [x]  when a block is given when self is a hash behaves like #each calls the given block once for each element in self. 

### my_select 
- [x] returns an array containing all elements of enum for which the given block returns a true value
- [x]  does not mutate the original array

### my_all? 
- [x] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

 ### my_any? 
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

### my_none? 
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

### my_inject 
- [ ] raises a "LocalJumpError" when no block or argument is given
- [ ] when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
- [ ]  when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- [ ] when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- [ ]  when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

**If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

File: enums.rb (hunk 2 above)
Comment (wajahat414) [reply enums.rb]: E: Lint/Syntax: unexpected token kEND
(Using Ruby 2.4 parser; configure using TargetRubyVersion parameter, under AllCops)
  end
can you telll why this error is showing?

**Review by SajjadAhmad14 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:

## Congratulations 🎉 Thank you for implementing changes suggested previously 🙏

### Happy Coding!

## You can always connect with me on slack ( @Sajjad Ahmad ).

**Pull request merged**